[Merged by Bors] - feat(data/fintype/order): complete order instances for fintype

apnelson1 opened the pull request:
This PR adds several instances (as defs) for fintypes:
* `order_bot` from `semilattice_inf`, `order_top` from `semilattice_sup`, `bounded_order` from `lattice`.
* `complete_lattice` from `lattice`.
* `complete_linear_order` from `linear_order`.

We use this last one to give a `complete_linear_order` instance for `fin (n + 1)` .

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

b-mehta commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,107 @@
+  le_top := λ a, fintype.le_sup_id (set.mem_univ a),
+  Inf := λ s, Sup {a | ∀ x ∈ s, a ≤ x},
+  Inf_le := λ a s h, fintype.sup_id_le (λ b hb, hb _ h),
+  le_Inf := λ a s h, fintype.le_sup_id (λ x hx, h _ hx)}
```
```suggestion

noncomputable instance fintype.has_Sup [fintype α] [semilattice_sup_bot α] :
  complete_semilattice_Sup α :=
{ Sup := λ S, S.to_finset.sup id,
  le_Sup := λ _ _, fintype.le_sup_id,
  Sup_le := λ _ _, fintype.sup_id_le,
  ..(infer_instance : semilattice_sup_bot α) }

@[priority 100] noncomputable instance fintype.complete_lattice_of_semilattice_sup_bot
[fintype α] [semilattice_sup_bot α]:
  complete_lattice α :=
complete_lattice_of_complete_semilattice_Sup _
```

This can be simplified drastically: the idea is that things like `le_antisymm` can be inferred from the existing partial order, and your definitions of Inf, inf are the same as the generic ones provided by `complete_lattice_of_complete_semilattice_Sup`. That docstring mentions that they're not the nicest definitionally, but the ones you provide are the same anyway, so it's not worse.

b-mehta commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,107 @@
+  sup := max,
+  le_sup_left := le_max_left,
+  le_sup_right := le_max_right,
+  sup_le := λ _ _ _ h h', max_le h h' }
```
```suggestion
@[priority 100] noncomputable instance fintype.semilattice_sup_bot_of_linear_order
[fintype α] [nonempty α] [linear_order α] :
  semilattice_sup_bot α :=
{ bot := classical.some (@fintype.exists_max α _ _ α (order_dual.linear_order α) id),
  bot_le := λ a, classical.some_spec (@fintype.exists_max α _ _ α (order_dual.linear_order α) id) _,
  sup := max,
  le_sup_left := le_max_left,
  le_sup_right := le_max_right,
  sup_le := λ _ _ _ h h', max_le h h',
  ..(infer_instance : linear_order α) }
```

As above, can condense

b-mehta commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,107 @@
+  le_total := linear_order.le_total,
+  decidable_le := infer_instance,
+  decidable_eq := infer_instance,
+  decidable_lt := infer_instance }
```
```suggestion
{ ..fintype.complete_lattice_of_semilattice_sup_bot,
  ..(infer_instance : linear_order α) }
```

b-mehta commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,107 @@
+@[priority 100] noncomputable instance fintype.semilattice_sup_bot_of_linear_order
+[fintype α] [nonempty α] [linear_order α] :
+  semilattice_sup_bot α :=
+{ bot := classical.some (@fintype.exists_max α _ _ α (order_dual.linear_order α) id),
```
This suggests that `fintype.exists_min` is missing!

b-mehta replied:
I've added it in #7126

b-mehta committed:
golf, move and dualise proof

apnelson1 committed:
incorporated exists_min

hrmacbeth commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,50 @@
+[fintype α] [nonempty α] [linear_order α] :
+  semilattice_sup_bot α :=
+{ bot := classical.some (fintype.exists_min id),
+  bot_le := λ a, classical.some_spec (fintype.exists_min id) _,
```
```suggestion
  bot_le := classical.some_spec (fintype.exists_min id),
```

hrmacbeth commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,50 @@
+complete_lattice_of_complete_semilattice_Sup _
+
+@[priority 100] noncomputable instance fintype.semilattice_sup_bot_of_linear_order
+[fintype α] [nonempty α] [linear_order α] :
```
```suggestion
  [fintype α] [nonempty α] [linear_order α] :
```

hrmacbeth commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,50 @@
+  ..(infer_instance : semilattice_sup_bot α) }
+
+@[priority 100] noncomputable instance fintype.complete_lattice_of_semilattice_sup_bot
+[fintype α] [semilattice_sup_bot α]:
```
```suggestion
  [fintype α] [semilattice_sup_bot α] :
```

hrmacbeth commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,50 @@
+  sup := max,
+  le_sup_left := le_max_left,
+  le_sup_right := le_max_right,
+  sup_le := λ _ _ _ h h', max_le h h',
```
```suggestion
  sup_le := λ _ _ _, max_le,
```

hrmacbeth reviewed (commented):
silly tweaks :)

apnelson1 committed:
Update src/data/fintype/order.lean

Co-authored-by: hrmacbeth <25316162+hrmacbeth@users.noreply.github.com>

apnelson1 committed:
Update src/data/fintype/order.lean

Co-authored-by: hrmacbeth <25316162+hrmacbeth@users.noreply.github.com>

apnelson1 committed:
Update src/data/fintype/order.lean

Co-authored-by: hrmacbeth <25316162+hrmacbeth@users.noreply.github.com>

apnelson1 committed:
Update src/data/fintype/order.lean

Co-authored-by: hrmacbeth <25316162+hrmacbeth@users.noreply.github.com>

eric-wieser commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,50 @@
+
+variables {α : Type*}
+
+@[priority 100] noncomputable instance fintype.has_Sup [fintype α] [semilattice_sup_bot α] :
```
What is it that makes this noncomputable?

b-mehta replied:
The same reason as in the module doc, it's the `to_finset` part.

eric-wieser replied:
Isn't `to_finset` computable though?

apnelson1 replied:
It relies on `subtype.fintype`, which needs a `decidable_pred` instance for the set in question. 

Mario says in this post that noncomputability is fine here. https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/fin.20n.20has_Inf.20and.20has_Sup/near/206633448

eric-wieser replied:
Capturing as much as that wisdom as possible in the module doc would be great. I see the point though, you'd need `decidable_pred (\mem s)` for all `s`, which is only possible with the axiom of choice

apnelson1 replied:
Thanks - updated the docstring.

b-mehta replied:
For the record, `to_finset` is noncomputable, as in https://github.com/leanprover-community/mathlib/blob/aa1fa0b2ba0739deaec3fe65f42640998120bc80/src/data/set/finite.lean#L36.

eric-wieser replied:
@b-mehta, that's `set.finite.to_finset`, not the `set.to_finset` relevant to this PR:

https://github.com/leanprover-community/mathlib/blob/aa1fa0b2ba0739deaec3fe65f42640998120bc80/src/data/fintype/basic.lean#L413-L416

b-mehta replied:
Good point, my mistake

b-mehta commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,50 @@
+
+variables {α : Type*}
+
+@[priority 100] noncomputable instance fintype.has_Sup [fintype α] [semilattice_sup_bot α] :
```
```suggestion
@[priority 100] noncomputable instance fintype.complete_semilattice_Sup [fintype α] [semilattice_sup_bot α] :
```
(This was purely my fault, sorry!)

apnelson1 committed:
docstring

apnelson1 committed:
fixed instance name

apnelson1 committed:
whitespace

gebner commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,52 @@
+
+variables {α : Type*}
+
+@[priority 100] noncomputable instance fintype.complete_semilattice_Sup
```
```suggestion
noncomputable def fintype.complete_semilattice_Sup
```
And similarly for the others, they all cause loops in type class search (because it goes from `linear_order α` to `complete_linear_order α` and then to `linear_order α` back again):
```lean
example : semilattice_sup_bot α := by apply_instance
/-
maximum class-instance resolution depth has been reached (the limit can be increased by setting option 'class.instance_max_depth') (the class-instance resolution trace can be visualized by setting option 'trace.class_instances')
state:
α : Type u_1
⊢ semilattice_sup_bot α
-/
```

apnelson1 committed:
changed instances to defs to avoid instance search problems
 added explicit instance for fin n+1

eric-wieser commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,56 @@
+  le_sup_left := le_max_left,
+  le_sup_right := le_max_right,
+  sup_le := λ _ _ _, max_le,
+  ..(infer_instance : linear_order α) }
```
Untested, but I think this works:
```suggestion
  ..(infer_instance : semilattice_sup α) }
```

eric-wieser commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,56 @@
+noncomputable def fintype.complete_lattice_of_semilattice_sup_bot (α : Type*)
+  [fintype α] [semilattice_sup_bot α] :
+  complete_lattice α :=
+@complete_lattice_of_complete_semilattice_Sup α (fintype.complete_semilattice_Sup α)
```
This definition feels like a bad idea to me - why not just provide `inf` explicitly as `s.to_finset.inf id?

apnelson1 replied:
I don't understand what you mean here.

eric-wieser replied:
I think you should add a matching `fintype.complete_semilattice_Inf` swapping `finset.sup` for `finset.inf`, then instead implement:

```suggestion
noncomputable def fintype.complete_lattice (α : Type*)
  [fintype α] [bounded_lattice α] : complete_lattice α :=
{ ..fintype.complete_semilattice_Sup,
  ..fintype.complete_semilattice_Inf }
```

That way you ensure that the `inf` of your resulting lattice is consistent with any `inf` that is already in place.

apnelson1 replied:
You need `bounded_lattice` for the second instance here, not just `lattice`. This raises the question of where to start - 
for a `fintype`,  any of `semilattice_sup_bot`, `semilattice_inf_top` and `lattice` are the same, and all give rise to `complete_lattice`. Would it make sense to have ways up to `complete_lattice` for all three?

eric-wieser replied:
> You need bounded_lattice for the second instance here, not just lattice.

Good catch

> Would it make sense to have ways up to complete_lattice for all three?

`complete_lattice_of_complete_semilattice_Sup` comes with some warnings indicating that it provides bad definitional equality. I'd be inclined to avoid it entirely - if the user wants it, they can apply it to `fintype.complete_semilattice_Sup` manually.

So I would probably lean towards providing the path to `complete_lattice` only for `bounded_lattice`.

apnelson1 replied:
I've just pushed a substantial rewrite which gets rid of `complete_semilattice_Sup` (/Inf) entirely, and instead just builds a bounded lattice then a complete lattice directly. This makes more of it computable.

eric-wieser commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,56 @@
+  complete_linear_order α :=
+let i := fintype.semilattice_sup_bot_of_linear_order α in
+{ ..@fintype.complete_lattice_of_semilattice_sup_bot α _ i,
+  ..(infer_instance : linear_order α) }
```
```suggestion
  ..(infer_instance : lattice α) }
```
would fill out more fields I think

apnelson1 replied:
This doesn't provide the `decidable`... fields.

eric-wieser commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,56 @@
+{ Sup := λ s, s.to_finset.sup id,
+  le_Sup := λ _ _ ha, finset.le_sup (set.mem_to_finset.mpr ha),
+  Sup_le := λ _ _ ha, finset.sup_le (λ b hb, ha _ (by rwa set.mem_to_finset at hb)),
+  ..(infer_instance : semilattice_sup_bot α) }
```
You should add a lemma that unfolds this `Sup` definition,
```lean
section

-- only for this lemma
local attribute [instance] fintype.complete_semilattice_Sup
 
/-- The `Sup` induced by `fintype.complete_semilattice_Sup` unfolds to `finset.sup`. -/
lemma fintype.Sup_eq {α : Type*} [fintype α] [semilattice_sup_bot α] (s : set α)
  [decidable_pred (∈ s)] : Sup s = s.to_finset.sup id :=
by { dsimp only [fintype.complete_semilattice_Sup, Sup], congr }

end
```

apnelson1 committed:
substantial rewrite -
lattice -> bounded_lattice -> complete_lattice
instead of going through complete_semilattice_Sup

apnelson1 committed:
added computable stuff

apnelson1 committed:
eq_top_iff

apnelson1 committed:
docstrings

apnelson1 committed:
fixed sections

b-mehta commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,110 @@
+  top := fintype.top α,
+  le_top := fintype.le_top α,
+  .. (infer_instance : lattice α)}
+
```
Right now there are two ways to refer to the top element of a fintype with lattice structure: `fintype.top` or `⊤`: I think there should be a simp lemma saying `fintype.top = ⊤` - that way things like `eq_top_iff` follow from general results on bounded lattices.

apnelson1 replied:
Thanks. This made me realize that defining `top`/`bot` outside structures didn't make sense in the first place. Latest push just upgrades `semilattice_sup` to `semilattice_sup_top` directly to gain access to `top` notation.

apnelson1 committed:
simplified finding top/bot

apnelson1 committed:
docstring typo

apnelson1 committed:
whitespace

b-mehta commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,91 @@
+{ top := top,
+  le_top := λ a, (((@finset.fold_op_rel_iff_and _ _ (⊔) _ _ _ (arbitrary α) _ (λ x y, y ≤ x)
+  (λ _ _ _, sup_le_iff) top )).mp le_rfl).2 a (finset.mem_univ a),
+  .. (infer_instance : semilattice_sup α)}
```
I think we can show the existence of `inf` on a fintype given `semilattice_sup_top`, right? Perhaps these constructions belong in this file; though I'm happy for them to be later PRs rather than making this one any bigger.

apnelson1 replied:
I don't think so - for instance, the divisibility semilattice on {2,3,6} has sup and top but no inf or bot.

b-mehta replied:
You're right, my mistake.

eric-wieser commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,91 @@
+  Inf_le := λ _ _ ha, @finset.inf_le _ _ iit _ _ _ (set.mem_to_finset.mpr ha),
+  le_Inf := λ s _ ha, @finset.le_inf _ _ iit s.to_finset _ _
+    (λ b hb, ha _ (by rwa set.mem_to_finset at hb)),
+  ..(@fintype.bounded_lattice α ⟨i.some⟩ _ _)}
```
For better control of definitional reduction of `inf` and `sup`:
```suggestion
/-- A finite bounded lattice is complete -/
noncomputable def fintype.complete_lattice (α : Type*) [fintype α] [h : bounded_lattice α] :
  complete_lattice α :=
let isb := (@semilattice_sup_bot_of_bounded_lattice α h),
    iit := (@semilattice_inf_top_of_bounded_lattice α h) in
{ Sup := λ s, @finset.sup _ _ isb s.to_finset id,
  Inf := λ s, @finset.inf _ _ iit s.to_finset id,
  le_Sup := λ _ _ ha, @finset.le_sup _ _ isb _ _ _ (set.mem_to_finset.mpr ha),
  Sup_le := λ s _ ha, @finset.sup_le _ _ isb s.to_finset _ _
    (λ b hb, ha _ (by rwa set.mem_to_finset at hb)),
  Inf_le := λ _ _ ha, @finset.inf_le _ _ iit _ _ _ (set.mem_to_finset.mpr ha),
  le_Inf := λ s _ ha, @finset.le_inf _ _ iit s.to_finset _ _
    (λ b hb, ha _ (by rwa set.mem_to_finset at hb)),
  ..h}
```

apnelson1 replied:
Would you advocate combining this with 
``noncomputable def fintype.complete_lattice_of_lattice (α : Type*) [fintype α] [i : nonempty α]
  [lattice α] : complete_lattice α :=
@fintype.complete_lattice _ _ (@fintype.bounded_lattice α ⟨i.some⟩ _ _)`` ?
I think it is still nice to be able to upgrade a finite `lattice` directly to a `complete_lattice`.

eric-wieser replied:
If you think `fintype.complete_lattice_of_lattice` is worth adding feel free to add it.
It's only in situations where a `bounded_lattice` is already available that it would be a bad idea to use it.

eric-wieser replied:
Of course, this doesn't scale well - if you have a chain of `n+1` typeclasses each stronger than the last, you end up with `n(n+1)/2` convenience defs.

eric-wieser commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,91 @@
+  .. (infer_instance : linear_order α) }
+
+noncomputable instance {n : ℕ} : complete_linear_order (fin (n+1)) :=
+  fintype.complete_linear_order  _
```
Note that my goal above is to ensure that this instance does not create any non-defeq diamonds - as written, `top = fin.last n` won't be true by definition.

eric-wieser commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,91 @@
+/-- The `Inf` induced by `fintype.complete_semilattice` unfolds to `finset.inf`. -/
+lemma fintype.Inf_eq {α : Type*} [nonempty α] [fintype α] [lattice α] (s : set α)
+  [decidable_pred (∈ s)] : Inf s = s.to_finset.inf id :=
+by {convert rfl}
```
I think it makes sense to have these for `top` and `bot` too, about `fintype.semilattice_sup_top` etc

apnelson1 committed:
rejigged for better defeq

apnelson1 committed:
typo

eric-wieser commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,119 @@
+  Inf_le := λ _ _ ha, @finset.inf_le _ _ iit _ _ _ (set.mem_to_finset.mpr ha),
+  le_Inf := λ s _ ha, @finset.le_inf _ _ iit s.to_finset _ _
+    (λ b hb, ha _ (by rwa set.mem_to_finset at hb)),
+  ..h}
```
You're making things hard for yourself here:

```suggestion
{ Sup := λ s, s.to_finset.sup id,
  Inf := λ s, s.to_finset.inf id,
  le_Sup := λ _ _ ha, finset.le_sup (set.mem_to_finset.mpr ha),
  Sup_le := λ s _ ha, finset.sup_le (λ b hb, ha _ (set.mem_to_finset.mp hb)),
  Inf_le := λ _ _ ha, finset.inf_le (set.mem_to_finset.mpr ha),
  le_Inf := λ s _ ha, finset.le_inf (λ b hb, ha _ (set.mem_to_finset.mp hb)),
  ..h}
```
`semilattice_inf_top_of_bounded_lattice` appears to be found just fine by typeclass search

apnelson1 committed:
Update src/data/fintype/order.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

eric-wieser commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,115 @@
+  [lattice α] : complete_lattice α :=
+@fintype.complete_lattice _ _ (@fintype.bounded_lattice α ⟨i.some⟩ _ _)
+
+/-- A nonempty finite linear order is complete -/
```
```suggestion
/-- A nonempty finite linear order is complete.

If the lattice is already a `bounded_lattice`, then build this object manually instead by combining
`fintype.complete_lattice` and an appropriate `linear_order` instance, as this gives definitional
equality for `⊥` and `⊤` -/
```

eric-wieser committed:
Merge branch 'master' into fintype_complete_linear_order

eric-wieser committed:
Update src/data/fintype/order.lean

eric-wieser commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,119 @@
+/-- The `Inf` induced by `fintype.complete_semilattice` unfolds to `finset.inf`. -/
+lemma fintype.Inf_eq {α : Type*} [fintype α] [bounded_lattice α] (s : set α)
+  [decidable_pred (∈ s)] : Inf s = s.to_finset.inf id :=
+by {convert rfl}
```
```suggestion
/-- The `Sup` induced by `fintype.complete_lattice` unfolds to `finset.sup`. -/
lemma fintype.Sup_eq {α : Type*} [fintype α] [bounded_lattice α] (s : set α)
  [decidable_pred (∈ s)] : Sup s = s.to_finset.sup id :=
by convert rfl

/-- The `Inf` induced by `fintype.complete_lattice` unfolds to `finset.inf`. -/
lemma fintype.Inf_eq {α : Type*} [fintype α] [bounded_lattice α] (s : set α)
  [decidable_pred (∈ s)] : Inf s = s.to_finset.inf id :=
by convert rfl
```

eric-wieser committed:
Update src/data/fintype/order.lean

eric-wieser commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,119 @@
+{ bot := bot,
+  bot_le := λ a, ((@finset.fold_op_rel_iff_and _ _ (⊓) _ _ _ _ _ _
+    (λ _ _ _, le_inf_iff) bot).mp le_rfl).2 a (finset.mem_univ _),
+  .. (infer_instance : semilattice_inf α)}
```
Once the build finishes I plan to come back to this and see if I can golf away the `@`

jcommelin replied:
@eric-wieser you still want to try this?

eric-wieser replied:
Slightly better:
```suggestion
{ bot := finset.fold (⊓) (arbitrary α) id finset.univ,
  bot_le := λ a, ((finset.fold_op_rel_iff_and (@le_inf_iff α _)).mp le_rfl).2 a (finset.mem_univ _),
  .. ‹semilattice_inf α› }
```

jcommelin commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,119 @@
+
+local attribute [instance] fintype.semilattice_inf_bot
+
+lemma fintype.bot_eq {α : Type*} [inhabited α] [fintype α] [semilattice_inf α] :
```
If you swap the two sides, then this could also be a `@[simp]` lemma.

jcommelin commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,119 @@
+
+local attribute [instance] fintype.semilattice_sup_top
+
+lemma fintype.top_eq {α : Type*} [inhabited α] [fintype α] [semilattice_sup α] :
```
Idem.

jcommelin reviewed (approved):
LGTM

eric-wieser commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,119 @@
+{ top := top,
+  le_top := λ a, (((@finset.fold_op_rel_iff_and _ _ (⊔) _ _ _ (arbitrary α) _ (λ x y, y ≤ x)
+  (λ _ _ _, sup_le_iff) top )).mp le_rfl).2 a (finset.mem_univ a),
+  .. (infer_instance : semilattice_sup α)}
```
The show is ugly, but perhaps less ugly than the `let` was:
```suggestion
{ top := finset.fold (⊔) (arbitrary α) id finset.univ,
  le_top := λ a,
    ((finset.fold_op_rel_iff_and (λ x y z, show x ≥ y ⊔ z ↔ _, from sup_le_iff)).mp le_rfl).2
      a (finset.mem_univ a),
  .. ‹semilattice_sup α› }
```

eric-wieser commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,119 @@
+/-- The `Inf` induced by `fintype.complete_lattice` unfolds to `finset.inf`. -/
+lemma fintype.Inf_eq {α : Type*} [fintype α] [bounded_lattice α] (s : set α)
+  [decidable_pred (∈ s)] : Inf s = s.to_finset.inf id :=
+by convert rfl
```
It would be nice if we could also prove:
```lean
/-- The `supr` induced by `fintype.complete_lattice` unfolds to `finset.sup`. -/
lemma fintype.supr_eq {α ι: Type*} [fintype ι] [fintype α] [bounded_lattice α] (f : ι → α) :
  supr f = finset.univ.sup f :=
begin
  classical,
  rw [supr, fintype.Sup_eq],
  have : (set.range f).to_finset = finset.univ.image f,
  { ext, simp?, },
  rw this,
  sorry,
end
```

eric-wieser replied:
I need #7428 to prove this

eric-wieser replied:
And also probably #7426 to avoid that `have`

eric-wieser committed:
Apply suggestions from code review

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into fintype_complete_linear_order

eric-wieser commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,117 @@
+
+import order.lattice
+import data.fintype.basic
+import order.category.NonemptyFinLinOrd
```
What's this import for?

eric-wieser commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,117 @@
+import order.category.NonemptyFinLinOrd
+
+/-!
+# Order structures on finite sets
```
```suggestion
# Order structures on finite types
```

eric-wieser committed:
add fintype.(supr|infi)_eq

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into fintype_complete_linear_order

YaelDillies committed:
bump mathlib

YaelDillies committed:
fix, add fin lemmas

YaelDillies committed:
simping

eric-wieser commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,143 @@
+by convert rfl
+
+/-- The `supr` induced by `fintype.to_complete_lattice` unfolds to `finset.sup`. -/
+lemma fintype.supr_eq (f : ι → α) : supr f = finset.univ.sup f :=
```
Don't we already have this lemma as `finset.univ_sup_eq_supr` or similar?

eric-wieser reviewed (commented):
Don't all the lemmas here hold generally for any complete lattice on a fintype, not just these ones?

eric-wieser commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,143 @@
+local attribute [instance] fintype.to_complete_lattice
+
+/-- The `Sup` induced by `fintype.to_complete_lattice` unfolds to `finset.sup`. -/
+lemma fintype.Sup_eq (s : set α) [decidable_pred (∈ s)] : Sup s = s.to_finset.sup id :=
```
This is almost `finset.sup_id_eq_Sup`,

YaelDillies committed:
trim off lemmas

urkud commented on src/data/fintype/order.lean:
```diff
@@ -0,0 +1,88 @@
+equality for `⊥` and `⊤`. -/
+noncomputable def fintype.to_complete_linear_order [h : linear_order α] : complete_linear_order α :=
+{ .. fintype.to_complete_lattice_of_lattice α,
+  .. h }
```
Could you please create two versions of this def (one with `[bounded_order α]` and one without)? Otherwise API for `complete_lattice` is inconsistent with API for `complete_order`.
Otherwise LGTM. Thanks!
bors d+

YaelDillies replied:
You might want to delegate me instead. I'm not sure Peter is checking much on mathlib these days.

YaelDillies committed:
add bounded_order version, reducible

YaelDillies committed:
fix unused argument

jcommelin reviewed (approved):
Thanks :tada:

bors merge
